File: app/schemas/application.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

APPLICATION_STATUSES = Literal["applied", "screening", "interview", "offer", "hired", "rejected"]


class StatusUpdate(BaseModel):
    status: APPLICATION_STATUSES
    notes: str | None = Field(default=None, max_length=2000)


class ApplicationOut(BaseModel):
    id: uuid.UUID
    job_id: uuid.UUID
    candidate_id: uuid.UUID
    document_id: uuid.UUID
    current_status: str
    applied_at: datetime

    class Config:
        from_attributes = True


class StatusHistoryOut(BaseModel):
    status: str
    changed_by: uuid.UUID
    changed_at: datetime
    notes: str | None

    class Config:
        from_attributes = True


class NoteCreate(BaseModel):
    content: str = Field(max_length=5000)


class NoteOut(BaseModel):
    id: uuid.UUID
    content: str
    created_by: uuid.UUID
    created_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/types/application.ts
export type ApplicationStatus =
  | 'applied'
  | 'screening'
  | 'interview'
  | 'offer'
  | 'hired'
  | 'rejected'

export interface ApplicationOut {
  id: string
  job_id: string
  candidate_id: string
  document_id: string
  current_status: ApplicationStatus
  applied_at: string
}

export interface StatusUpdate {
  status: ApplicationStatus
  notes?: string | null
}

export interface StatusHistoryOut {
  status: ApplicationStatus
  changed_by: string
  changed_at: string
  notes: string | null
}

export interface ApplicationDetail {
  id: string
  job: {
    id: string
    title: string
    company: string
    job_type: string
  }
  candidate: {
    id: string
    name: string
  }
  current_status: ApplicationStatus
  applied_at: string
  interview_rounds: {
    round_number: number
    outcome: string
    scheduled_at: string | null
  }[]
  status_history: {
    status: ApplicationStatus
    changed_at: string
  }[]
}

export interface NoteCreate {
  content: string
}

export interface NoteOut {
  id: string
  content: string
  created_by: string
  created_at: string
}

export interface ResumeDownload {
  url?: string
  filename: string
  expires_in_seconds?: number
  expires_in_human?: string
}
